Packages/com.unity.inputsystem/InputSystem/SourceGenerators/Source~/SourceGenerators/GlobalInputActionsSourceGenerator.cs

```diff
@@ -0,0 +1,372 @@
+        }
+
+        private static string GenerateInputActionMapClassName(ActionMap actionMap)
+        {
+            // TODO: More robust class name generation. Replace incompatible characters
+            var actionMapClassName = actionMap.Name.Replace(" ", "");
+
+            return FormatClassName(actionMapClassName) + "InputActionMap";
+        }
+
+        private static string FormatClassName(string str)
+        {
+            return "_" + char.ToUpper(str[0]) + str.Substring(1);
+        }
+
+        private static string FormatFieldName(string str)
+        {
+            if (str.Length <= 3)
+                return (char.IsDigit(str[0]) ? "_" : "") + str.ToUpper();
+
+            return (char.IsDigit(str[0]) ? "_" : "") + char.ToLower(str[0]) + str.Substring(1);
+        }
+
+        private string GetTypeFromExpectedType(ControlType controlType)
+        {
+            switch (controlType)
+            {
+                case ControlType.Analog:
+                case ControlType.Axis:
+                case ControlType.Button:
+                case ControlType.Delta:
+                case ControlType.DiscreteButton:
+                case ControlType.Key:
+                    return nameof(Single);
+
+                case ControlType.Digital:
+                case ControlType.Integer:
+                    return nameof(Int32);
+
+                case ControlType.Double:
+                    return nameof(Double);
+
+                case ControlType.Bone:
+                    return "UnityEngine.InputSystem.XR.Bone";
+
+                case ControlType.Dpad:
+                case ControlType.Vector2:
+                case ControlType.Stick:
+                    return "UnityEngine.Vector2";
+
+                case ControlType.Vector3:
+                    return "UnityEngine.Vector3";
+
+                case ControlType.Eyes:
+                    return "UnityEngine.InputSystem.XR.Eyes";
+
+                case ControlType.Pose:
+                    return "UnityEngine.InputSystem.XR.PoseState";
+
+                case ControlType.Quaternion:
+                    return "UnityEngine.Quaternion";
+
+                case ControlType.Touch:
+                    return "UnityEngine.InputSystem.LowLevel.TouchState";
+
+                default:
+                    return null;
+            }
+        }
+
+        public static int GetStableHashCode(string str)
+        {
+            unchecked
+            {
+                var hash1 = 5381;
+                var hash2 = hash1;
+
+                for (var i = 0; i < str.Length && str[i] != '\0'; i += 2)
+                {
+                    hash1 = ((hash1 << 5) + hash1) ^ str[i];
+                    if (i == str.Length - 1 || str[i + 1] == '\0')
+                        break;
+                    hash2 = ((hash2 << 5) + hash2) ^ str[i + 1];
+                }
+
+                return hash1 + (hash2 * 1566083941);
+            }
+        }
+    }
+
+    public struct InputActionAsset
```

Comment:
Can we share any of these definitions with the main package code to avoid duplication and risk of going out of sync.
Or at least comment both versions with a reference of where to update to its clear they need to be kept in sync.

Reply:
They can't be shared ASAIK as this is not a Unity project but a MSBuild project (maybe neutron will change this story).
But yeah, agree it needs to be documented as there needs to be parity between the asset and the generator code.